### PR TITLE
Active Bounty Refactor

### PR DESCRIPTION
**Pull request by m1sterc001guy**
Refactors the voting logic to support Active Bounties. Bounties are no longer required to be accepted before voting can take place. The "Active" state is effectively the same as the "Open" state, but we will display it differently on the front end. This means that active bounties can also be expired. 

New timestamp was introduced on the contracts table that records when the contract transitions back to the "open" state.

### CODE REVIEW

File: api/src/main/java/com/nicknathanjustin/streamercontracts/contracts/dtos/PrivateContract.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+@Data
+public class PrivateContract extends Contract {
+
+    private boolean userHasVoted;
+    
+    public PrivateContract(@NonNull final ContractModel contract) {
+        super(contract);
+        this.userHasVoted = false;
+        for (VoteModel vote : contract.getVotes()) {
```
Comment (nckackerman) [inline api/src/main/java/com/nicknathanjustin/streamercontracts/contracts/dtos/PrivateContract.java]: Lambdas/Streams are the cool new java8 way to do loops:
```
contract.getVotes().stream().filter(vote -> vote.getVoter().getTwitchUsername().equals(contract.getStreamer().getTwitchUsername())))
```
^^filters the list done to votes where the voter is the streamer. If the list is greater than 0 we know the streamer voted... Disclaimer: I didnt test to see if this code compiles or works correctly.

File: api/src/main/java/com/nicknathanjustin/streamercontracts/votes/VoteServiceImpl.java [hunk 2]
```diff
@@ -22,7 +22,10 @@
     public void recordVote(@NonNull final UserModel voter,
                            @NonNull final ContractModel contractModel,
                            final boolean flaggedCompleted) {
-        assertContractIsVotable(contractModel);
+    	if (isContractPastSettleTimestamp(contractModel)) {
```
Comment (nckackerman) [inline api/src/main/java/com/nicknathanjustin/streamercontracts/votes/VoteServiceImpl.java]: Is this check better placed in the assertUserCanVoteOnContract() method?

File: api/src/main/java/com/nicknathanjustin/streamercontracts/contracts/ContractService.java [hunk 3]
```diff
@@ -60,56 +61,60 @@
      * @param streamer The user to retrieve contracts by.
      * @param state The state to retrieve contracts by.
      * @param pageable identifies the page number and pagesize to retrieve
+     * @param isLoggedIn flag that indicates if the user is logged in.
      * @return a set of all contracts in the given state for the given user.
      */
-    Page<ContractDto> getContractsForStreamerAndState(UserModel streamer, ContractState state, Pageable pageable);
+    Page<Contract> getContractsForStreamerAndState(UserModel streamer, ContractState state, Pageable pageable, boolean isLoggedIn);
```
Comment (nckackerman) [inline api/src/main/java/com/nicknathanjustin/streamercontracts/contracts/ContractService.java]: when I read 'isLoggedIn' i wasnt sure if it was for any logged in user or only the loggedIn streamer... Im struggling to think of a better name. Maybe something like 'isPrivate' or 'requestedByStreamer' could be a little more clear.

A different option would be to pass the requester's twitchUsername to the service as a string. Then in the service method check if the requester is the contract's owner, and if it is, return the privateDto.

I favor passing the requester's twitchUsername the most... What do you think?

Comment (m1sterc001guy) [reply api/src/main/java/com/nicknathanjustin/streamercontracts/contracts/ContractService.java]: Little bit confused by "any logged in user or only the logged in streamer". They are the same thing right?

File: api/src/main/java/com/nicknathanjustin/streamercontracts/votes/VoteServiceImpl.java (hunk 2 above)
Comment (m1sterc001guy) [reply api/src/main/java/com/nicknathanjustin/streamercontracts/votes/VoteServiceImpl.java]: yea good call

**Pull request merged**